File: java/Relativedelta.java
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.HashMap;
 
public class Relativedelta {
 
    private Integer year = null;
    private Integer month = null;
    private Integer day = null;
    private Integer hour = null;
    private Integer minute = null;
    private Integer second = null;
 
    private int years = 0;
    private int months = 0;
    private int days = 0;
    private int hours = 0;
    private int minutes = 0;
    private int seconds = 0;
    private int leapdays = 0;
    private HashMap<String, Integer> map = null;
 
    public Relativedelta(LocalDateTime baseDay, LocalDateTime day) {
 
        if (day.compareTo(baseDay) == -1) {
            // 日付が基準日より前の場合は例外とする
            throw new IllegalArgumentException();
        }
 
        var months = (day.getYear() - baseDay.getYear()) * 12 + (day.getMonthValue() - baseDay.getMonthValue());
        this.setMonths(months);
 
        var dtm = this.getDTM(baseDay);
 
        var baseSec = baseDay.toEpochSecond(ZoneOffset.systemDefault().getRules().getOffset(baseDay));
        var daySec = day.toEpochSecond(ZoneOffset.systemDefault().getRules().getOffset(day));
 
        MyCompare compare = null;
        var increment = 0;
        if (daySec < baseSec) {
            compare = new MyCompare(true);
            increment = 1;
        } else {
            compare = new MyCompare(false);
            increment = -1;
        }
 
        while (compare.compare(day, dtm) == 1) {
            months += increment;
            this.setMonths(months);
            dtm = getDTM(baseDay);
        }
        var duration = Duration.between(dtm, day);
        this.seconds = (int) duration.toSeconds();
 
        if (this.seconds > 59) {
            var s = (int) Math.signum(this.seconds);
            var div = Math.floorDiv(this.seconds * s, 60);
            var mod = (int) (this.seconds * s) % 60;
            this.seconds = mod * s;
            this.minutes += div * s;
        }
        if (this.minutes > 59) {
            var s = (int) Math.signum(this.minutes);
            var div = Math.floorDiv(this.minutes * s, 60);
            var mod = (int) (this.minutes * s) % 60;
            this.minutes = mod * s;
            this.hours += div * s;
        }
        if (this.hours > 23) {
            var s = (int) Math.signum(this.hours);
            var div = Math.floorDiv(this.hours * s, 24);
            var mod = (int) (this.hours * s) % 24;
            this.hours = mod * s;
            this.days += div * s;
        }
        if (months > 11) {
            var s = (int) Math.signum(this.months);
            var div = Math.floorDiv(this.months * s, 12);
            var mod = (int) (this.months * s) % 12;
            this.months = mod * s;
            this.years += div * s;
        }
 
        var map = new HashMap<String, Integer>();
        map.put("y", this.years);
        map.put("m", this.months);
        map.put("d", this.days);
        map.put("hh", this.hours);
        map.put("mm", this.minutes);
        map.put("ss", this.seconds);
 
        this.map = map;
    }
 
    private void setMonths(int months) {
        this.months = months;
        if (this.months > 11) {
            var s = (int) Math.signum(this.months);
            var div = Math.floorDiv(this.months * s, 12);
            var mod = (int) (this.months * s) % 12;
            this.months = mod * s;
            this.years = div * s;
        } else {
            this.years = 0;
        }
    }
 
    private LocalDateTime getDTM(LocalDateTime other) {
        var y = 0;
        if (this.year != null) {
            y = this.year + this.years;
        } else if (other.getYear() > 0) {
            y = other.getYear() + this.years;
        }
        var m = this.month != null ? this.month : other.getMonthValue();
        if (this.months > 0) {
            m += this.months;
            if (m > 12) {
                y += 1;
                m -= 12;
            } else if (m < 1) {
                y -= 1;
                m += 12;
            }
        }
        var _d = this.day != null ? this.day : other.getDayOfMonth();
        var d = Math.min(LocalDateTime.of(y, m, 1, 0, 0, 0).toLocalDate().lengthOfMonth(), _d);
 
        var repl = new HashMap<String, Integer>();
        repl.put("year", y);
        repl.put("month", m);
        repl.put("day", d);
        repl.put("hour", this.hour != null ? this.hour : other.getHour());
        repl.put("minute", this.minute != null ? this.minute : other.getMinute());
        repl.put("second", this.second != null ? this.second : other.getSecond());
 
        var days = this.days;
        var ldt = LocalDateTime.of(y, 1, 1, 0, 0, 0);
        if (this.leapdays > 0 && m > 2 && ldt.toLocalDate().isLeapYear()) {
            days += this.leapdays;
        }
        var ret = LocalDateTime.of(
                repl.get("year"),
                repl.get("month"),
                repl.get("day"),
                repl.get("hour"),
                repl.get("minute"),
                repl.get("second"));
 
        return ret.plusDays(days).plusHours(this.hours).plusMinutes(this.minutes).plusSeconds(this.seconds);
    }
 
    public HashMap<String, Integer> getMap() {
        return this.map;
    }
 
    private class MyCompare implements java.util.Comparator<LocalDateTime> {
 
        private boolean isGT;
 
        public MyCompare(boolean isGT) {
            this.isGT = isGT;
        }
 
        @Override
        public int compare(LocalDateTime o1, LocalDateTime o2) {
            var o1Sec = o1.toEpochSecond(ZoneOffset.systemDefault().getRules().getOffset(o1));
            var o2Sec = o2.toEpochSecond(ZoneOffset.systemDefault().getRules().getOffset(o2));
 
            var ret = 0;
            if (isGT) {
                if (o1Sec > o2Sec)
                    ret = 1;
                else if (o1Sec < o2Sec)
                    ret = -1;
                return o1Sec > o2Sec ? -1 : 1;
            } else {
                if (o2Sec > o1Sec)
                    ret = 1;
                else if (o2Sec < o1Sec)
                    ret = -1;
            }
            return ret;
        }
 
    }
}

File: kotlin/Relativedelta.kt
import java.time.Duration
import java.time.LocalDateTime
import java.time.ZoneOffset
import kotlin.math.min
import kotlin.math.sign
 
 
internal class Relativedelta(baseDay: LocalDateTime, day: LocalDateTime) {
  private val year: Int? = null
  private val month: Int? = null
  private val day: Int? = null
  private val hour: Int? = null
  private val minute: Int? = null
  private val second: Int? = null
  private var years = 0
  private var months = 0
  private var days = 0
  private var hours = 0
  private var minutes = 0
  private var seconds = 0
  private val leapdays = 0
  private var map: HashMap<String, Int>? = null
 
  init {
    // 日付が基準日より前の場合は例外とする
    require(day.compareTo(baseDay) != -1)
    var months = (day.year - baseDay.year) * 12 + (day.monthValue - baseDay.monthValue)
    setMonths(months)
    val baseSec = baseDay.toEpochSecond(ZoneOffset.systemDefault().rules.getOffset(baseDay))
    val daySec = day.toEpochSecond(ZoneOffset.systemDefault().rules.getOffset(day))
    var dtm = getDTM(baseDay)
    var compare: MyCompare? = null
    var increment = 0
    if (daySec < baseSec) {
      compare = MyCompare(true)
      increment = 1
    } else {
      compare = MyCompare(false)
      increment = -1
    }
    while (compare.compare(day, dtm) == 1) {
      months += increment
      setMonths(months)
      dtm = getDTM(baseDay)
    }
    val duration = Duration.between(dtm, day)
    this.seconds = duration.toSeconds().toInt()
    if (this.seconds > 59) {
      val s = sign(this.seconds.toDouble()).toInt()
      val div = Math.floorDiv(this.seconds * s, 60)
      val mod = (this.seconds * s) % 60
      this.seconds = mod * s
      this.minutes += div * s
    }
    if (this.minutes > 59) {
      val s = sign(this.minutes.toDouble()).toInt()
      val div = Math.floorDiv(this.minutes * s, 60)
      val mod = (this.minutes * s) % 60
      this.minutes = mod * s
      this.hours += div * s
    }
    if (this.hours > 23) {
      val s = sign(this.hours.toDouble()).toInt()
      val div = Math.floorDiv(this.hours * s, 24)
      val mod = (this.hours * s) % 24
      this.hours = mod * s
      this.days += div * s
    }
    if (months > 11) {
      val s = sign(this.months.toDouble()).toInt()
      val div = Math.floorDiv(this.months * s, 12)
      val mod = (this.months * s) % 12
      this.months = mod * s
      this.years += div * s
    }
    val map = HashMap<String, Int>()
    map["y"] = this.years
    map["m"] = this.months
    map["d"] = this.days
    map["hh"] = this.hours
    map["mm"] = this.minutes
    map["ss"] = this.seconds
    this.map = map
  }
 
  private fun setMonths(months: Int) {
    this.months = months
    if (this.months > 11) {
      val s = sign(this.months.toDouble()).toInt()
      val div = Math.floorDiv(this.months * s, 12)
      val mod = (this.months * s) % 12
      this.months = mod * s
      this.years = div * s
    } else {
      this.years = 0
    }
  }
 
  private fun getDTM(other: LocalDateTime): LocalDateTime {
    var y = 0
    if (this.year != null) {
      y = this.year + this.years
    } else if (other.year > 0) {
      y = other.year + this.years
    }
    var m = this.month ?: other.monthValue
    if (this.months > 0) {
      m += this.months
      if (m > 12) {
        y += 1
        m -= 12
      } else if (m < 1) {
        y -= 1
        m += 12
      }
    }
    val _d = this.day ?: other.dayOfMonth
    val d =
      min(LocalDateTime.of(y, m, 1, 0, 0, 0).toLocalDate().lengthOfMonth().toDouble(), _d.toDouble())
        .toInt()
    val repl = HashMap<String, Int?>()
    repl["year"] = y
    repl["month"] = m
    repl["day"] = d
    repl["hour"] = this.hour ?: other.hour
    repl["minute"] = this.minute ?: other.minute
    repl["second"] = this.second ?: other.second
    var days = this.days
    val ldt = LocalDateTime.of(y, 1, 1, 0, 0, 0)
    if (this.leapdays > 0 && m > 2 && ldt.toLocalDate().isLeapYear) {
      days += this.leapdays
    }
    val ret = LocalDateTime.of(
      repl["year"]!!,
      repl["month"]!!,
      repl["day"]!!,
      repl["hour"]!!,
      repl["minute"]!!,
      repl["second"]!!
    )
    return ret.plusDays(days.toLong()).plusHours(this.hours.toLong()).plusMinutes(this.minutes.toLong())
      .plusSeconds(this.seconds.toLong())
  }
 
  fun getMap(): HashMap<String, Int>? {
    return this.map
  }
 
  private inner class MyCompare(private val isGT: Boolean) : Comparator<LocalDateTime> {
    override fun compare(o1: LocalDateTime, o2: LocalDateTime): Int {
      val o1Sec = o1.toEpochSecond(ZoneOffset.systemDefault().rules.getOffset(o1))
      val o2Sec = o2.toEpochSecond(ZoneOffset.systemDefault().rules.getOffset(o2))
      var ret = 0
      if (isGT) {
        if (o1Sec > o2Sec) ret = 1 else if (o1Sec < o2Sec) ret = -1
        return if (o1Sec > o2Sec) -1 else 1
      } else {
        if (o2Sec > o1Sec) ret = 1 else if (o2Sec < o1Sec) ret = -1
      }
      return ret
    }
  }
}